packet/src/pacer.rs

```diff
@@ -0,0 +1,131 @@
+use std::time::{Duration, Instant};
+
+pub struct LeakyBucketPacer {
+    /// Target bitrate in bits per second.
+    target_bitrate: u64,
+    /// The last time we refreshed the budget.
+    last_budget_increase: Option<Instant>,
+    /// The current budget in bytes.
+    budget: usize,
+    /// The current pacing i.e. how frequently we refresh the budget.
+    pacing: Duration,
```

Comment:
I admit I don't understand why this is needed.   Why not just refresh the budget with every packet?  That's what the libwebrtc pacer does.  Both it and the pacer I wrote work in terms of "last packet sent" and "next time to send a packet", which I think would avoid any delays introduced by this value.

https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/modules/pacing/pacing_controller.cc;l=306;drc=c14f4e9399fc4e23a6c4c68fc60086bdceec1c90;bpv=1;bpt=1

Reply:
The thinking here is that this will be used as follows

```rust
// On timeout
// Aim to call this every 5 ms
pacer.refresh_budget(now);
```

In [poll_packet](https://github.com/algesten/str0m/blob/6ebbc9776e8dc1b300092087d322b7f93b3f52c0/str0m/src/session.rs#L603)

```rust
        for m in only_media_mut(&mut self.media) {
            let twcc_seq = self.twcc;
            if let Some((header, seq_no)) = m.poll_packet(now, &self.exts, &mut self.twcc, buf) {
                trace!("Poll RTP: {:?}", header);

                let protected = srtp_tx.protect_rtp(buf, &header, *seq_no);
                let should_send = self
                    .pacer
                    .budget_allows_send(protected.len());

                if should_send {
                    self.twcc_tx_register.register_seq(twcc_seq.into(), now);
                    return Some(protected.into());
                } else {
                    self.held_srtp_packet = Some((protected, twcc_seq));

                    break;
                }
            }
        }

```

It's a bit different than if the queue lived within the pacer. The queued packets here are within each media line.

Str0m's contract is to call `poll_output` until a timeout is returned and here we'll return all packets allowed by the budget before doing so.

Reply:
I think the problem is just that 5ms is not high resolution enough, so it seems like it's better to just call poll_timeout() and have that return the time when the next packet should be sent, at which point you would call handle_timeout/poll_output.  That's why I think it's easier to think of this not as whether or not to send or hold a given packet, and instead to think of this as there's a time you want to send a packet.  Or, to use the term of the name of the struct, there is a time to leak the next drop out of the bucket.   And, yes, that's easier to do when you have the contents of the bucket.  

But if you didn't want to hold the queue in the pacer, I suppose what you could do is pass info about the queue (perhaps even an &mut) every time you call poll_output.  So it doesn't require holding the queue, but it doesn't require knowing something about the queue at the time poll_output is called.

Reply:
That all makes sense, except the singular use of *packet*, isn't it the next time when *packets* should be sent i.e. we queue up packets until the next send time and then send them all. My understanding is that this "burstiness" is important to establish and identify the packet groups to later calculate the inter group delay variation

Reply:
The entire point of a pacer is to _avoid_ burstiness.  And googcc does not rely on burstiness to function.  

You may be thinking of probing, which has specific bursts that are done at a higher rate than the target rate (thus decreasing the interval, or increasing the burstiness).  But even when probing, you're not increasing burstiness such that the packets are sent with _0_ interval between them.  It's just a probe with a decreased delay (an increased rate).  In fact, when probing, there's a specific rate/interval between them that is being probed.

For example, see here:

https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/modules/pacing/bitrate_prober.cc;l=178;drc=3297fb6d70a02d8c4af8db2bc9e315053744442b;bpv=1;bpt=1

Reply:
Hmm,

my understanding is based on section 4 of the IETF draft:

>  Pacing is used to actuate the target bitrate computed by the
   controllers.
>   When media encoder produces data, this is fed into a Pacer queue.
   The Pacer sends a group of packets to the network every burst_time
   interval.  RECOMMENDED value for burst_time is 5 ms.  The size of a
   group of packets is computed as the product between the target
   bitrate and the burst_time.

as well as the notion that forming these packet groups and identifying their receive deltas is important to calculate the inter group delay variation.

It sounds like this is one of those instances when the IETF draft is outdated then.


With the approach you outline and what libWebRTC does wouldn't knowledge of the buffers be required to make decisions around when the next send time should be meaning the proposed design here is no good?

Reply:
Interesting.  libwebrtc can send more than one packet a time if certain field trails are on (see here: https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/modules/pacing/task_queue_paced_sender.cc;l=81;drc=3297fb6d70a02d8c4af8db2bc9e315053744442b;bpv=1;bpt=1 and https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/modules/pacing/pacing_controller.cc;l=75;drc=3297fb6d70a02d8c4af8db2bc9e315053744442b;bpv=1;bpt=1)

I don't believe that's the default behavior.  The default appears to be a 0 for the burst interval, which means no bursting. But I could be wrong.  Maybe I'm missing something.  

I do see one place where a 5ms constant is used that has the comment "Time limit in milliseconds between packet bursts.", but that value is only used once: to give the initial budget size, which I think means the pacers acts as though 5ms has already passed when it starts.  

So, it's definitely possible I'm missing something fundamental.  The only way to get it right may be to read the libwebrtc code very, very carefully (perhaps more carefully than I thought I did already).

Reply:
The true value seems to be 40. It's [set](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/peerconnection/rtc_peer_connection_handler.cc;l=1089?q=pacer_burst_interval&ss=chromium) in Chromium an passed all the way down to `PacingController` where it's stored in `send_burst_interval_`. You can see how it's used in `PacingController::NextSendTime`.

It's strange that other parts of the code, to do with identifying inter arrival groups, still use the value 5ms